frontend/middle-3/fp.md

```diff
@@ -0,0 +1,183 @@
+* [Point-free gotchas in JavaScript](https://dev.to/danhomola/point-free-gotchas-in-javascript--3pfi)
+* [Thinking in Ramda: Pointfree Style](http://randycoulman.com/blog/2016/06/21/thinking-in-ramda-pointfree-style/)
+* [Achieving point-free JavaScript with R.converge / _.over](https://medium.com/@nicoespeon/achieving-point-free-javascript-with-r-converge-overargs-94060fd0ed7a)
+* [Why Ramda?](https://fr.umio.us/why-ramda/)
+
+
+## Immutability
+  * What is immutability? Can program with immutable data only be stateful? Is stateless programs are all from immutable data?
+  * Why is it dangerous to use mutable data structures as keys for hash maps?
+  * Why does mutability make caching hard?
+  * Why does mutability make concurrency hard?
+  * Why does mutability make inheritance hard? What is Liskov Substitution Principle and how mutability is related to this principle?
+  * Can mutability break encapsulation (if it is true, give examples)?
+  * How can we avoid state? When is state inevitable?
+  * What are best practices for localization of the state?
+  * What is an atomicity? How it can help to localize a state?
+  * How two-phase lifecycle can help to localize a state? What's the difference between static and dynamic two-phase lifecycle?
+  * What are clients of the state? Why its number should be as few as possible? Do clients have state itself?
+  * What are inconsistency of the state? How can we guarantee consistency of the state for outer client?
+  * How can dependency break help us to avoid mutable interface with a state?
+  * What is curried object? How can it help to deal with a problem of state?
+  * What are idempotency and commutativity? Why are they useful?
+  * How does Immutable.js work internally?
+  * Why does mutability tightly couple data with computations?
+  * What is the tradeoff between scalability and performance?
+  * Why do stateful applications scale worse than stateless?
+
+### Resources:
+* [On stateless software design](https://leonmergen.com/on-stateless-software-design-what-is-state-72b45b023ba2)
+* [Why We Need Explicit State](http://michaelrbernste.in/2013/12/10/why-we-need-explicit-state.html)
+* [Mutability Leads to Suffering](https://hackernoon.com/mutability-leads-to-suffering-23671a0def6a)
+* [Values and Change: Clojure’s approach to Identity and State](https://clojure.org/about/state)
+* [The discussion thread about mutability](http://lambda-the-ultimate.org/node/724#comment-6580)
+
+## Algebraic Data Types
+* What are Algebraic Data Types (AlgDT)?
+* What is the difference between Sum Types and Unions?
+* Why are Sums and Products called so?
+* Why AlgDT are called also inductive data types?
+* Why Sum Types are called also tagged union? What is a tag there?
+* How algebraic properties respected in AlgDT (like Symmetry, Identity, Associativity and Distributivity)?
+* What is a difference in tracking nullable values with Unions and with Sum Types?
```

Comment:
What is the difference between tracking nullable values with Unions and Sum Types?